### DISCUSSION TITLE
'I2C' object has no attribute 'try_lock'

### DISCUSSION THREAD

**Discussion by senaoezdogan** [micropython/micropython#12419] [RP2040 / Pico]
I am encountering an issue while working with the I2C in MicroPython.
I am using a KY-039 heart beat sensor with a KY-053 Analog-Digital Converter but it throws me following error: 'I2C' object has no attribute 'try_lock'

- **MicroPython Version**:  1.20.0
- **Hardware**: Raspberry-pi-pico
- **Code Snippet**:
```
import time
import machine
import adafruit_ads1x15.ads1115 as ADS
from adafruit_ads1x15.analog_in import AnalogIn

i2c = machine.I2C(0, scl=machine.Pin(1), sda=machine.Pin(0))

ads = ADS.ADS1115(i2c)

chan0 = AnalogIn(ads, ADS.P0)
chan1 = AnalogIn(ads, ADS.P1)
chan2 = AnalogIn(ads, ADS.P2)
chan3 = AnalogIn(ads, ADS.P3)

if __name__ == '__main__':
    GAIN = 2/3
    curState = 0
    thresh = 525  # mid point in the waveform
    P = 512
    T = 512
    stateChanged = 0
    sampleCounter = 0
    lastBeatTime = 0
    firstBeat = True
    secondBeat = False
    Pulse = False
    IBI = 600
    rate = [0]*10
    amp = 100

    lastTime = int(time.time()*1000)

    while True:
        # Read from the ADC
        Signal = chan0.value  # Modify this to read from the correct ADC channel

        curTime = int(time.time()*1000)
        sampleCounter += curTime - lastTime
        lastTime = curTime
        N = sampleCounter - lastBeatTime

        if Signal < thresh and N > (IBI/5.0)*3.0:
            if Signal < T:
                T = Signal

        if Signal > thresh and Signal > P:
            P = Signal

        if N > 250:
            if (Signal > thresh) and (Pulse == False) and (N > (IBI/5.0)*3.0):
                Pulse = True
                IBI = sampleCounter - lastBeatTime
                lastBeatTime = sampleCounter

                if secondBeat:
                    secondBeat = False
                    for i in range(0, 10):
                        rate[i] = IBI

                if firstBeat:
                    firstBeat = False
                    secondBeat = True
                    continue

                runningTotal = 0

                for i in range(0, 9):
                    rate[i] = rate[i+1]
                    runningTotal += rate[i]

                rate[9] = IBI
                runningTotal += rate[9]
                runningTotal /= 10
                BPM = 60000/runningTotal
                print('BPM: {}'.format(BPM))

        if Signal > thresh and Pulse == True:
            Pulse = False
            amp = P - T
            thresh = amp/2 + T
            P = thresh
            T = thresh

        if N > 2500:
            thresh = 512
            P = 512
            T = 512
            lastBeatTime = sampleCounter
            firstBeat = True
            secondBeat = False
            print("no beats found")

        time.sleep(0.005)
    ```

**Comment by robert-hh** [comment]
You cannot directly use CircuitPython drivers with MicroPython. As analternative, you might try this driver for ADS1115: https://github.com/robert-hh/ads1x15

**Reply by robert-hh** [reply]
Note: Please enclose code into lines with three backticks ```. Them the format is maintained. If the leading line is ```python, then the source is nicely colored.

**Comment by rkompass** [comment]
Circuitpython uses the same I2C object, but requires an additional locking step in the setup of the I2C bus.

Where we have in Micropython, e.g. on the RP2 Pico:

```py
from machine import I2C,  Pin          # Raspberry Pi Pico, Micropython
i2c = I2C(0, sda=Pin(0), scl=Pin(1), freq=100000)
```
In Circuitpython the equivalent is:
```py
import board                           # Raspberry Pi Pico, Circuitpython
from busio import I2C
i2c = I2C(sda=board.GP0, scl=board.GP1)
while i2c.try_lock():                  # CP requires I2C bus locking
    pass
```
Afterwards you may use the i2c object in similar way, i.e. with the same commands.

So, if changing for an MP driver for the ads1x15 seems too complicated, you also might try to find the above locking statement in the Circuitpython driver and remove it for the MP setup.

**Reply by dhalbert** [reply]
The `adafruit_bus_device` library, which wraps the `busio.I2C()` object, takes care of locking automatically when used with context managers. You can also use the Adafruit_Blinka library to emulate CircuitPython I/O on top of MicroPython: https://learn.adafruit.com/circuitpython-libraries-on-micropython-using-the-raspberry-pi-pico

**Reply by rkompass** [reply]
Hello Dan  (@dhalbert) ,

I have 2 questions:

1: Could you shortly explain why you require the i2c locking? What are the advantages / which problems are avoided by this?

2: I read about Blinka yesterday. But it was not clear for me whether Blinka may be used on Micropython Unix(Linux) to access i2c/spi peripherals via USB adapter boards. It it might, then that would be very useful. Does this possibility exist?

**Reply by dhalbert** [reply]
1. The locking is to insure that an I2C transaction completes atomically. I was not involved in that design. In practice, since we don't provide preemptive multitasking, this is unlikely, but the capability is there.

2. Blinka is mostly used to allow CircuitPython code to run on top of CPython on host computers, both SBC's like RPi, and with USB adapters. We haven't tried Blinka as a wrapper on top of Unix MicroPython, since using CPython is the easier use case.
